doc/design/persistent_volumes_etcd_data.md

```diff
@@ -0,0 +1,79 @@
+- If podPolicy has persistent volumes enabled, when creating a pod also create a Persistent Volume Claim and use it as the volume source for etcd-data
+- Handle etcd data persistent volume garbage collection like already done with other resources. While backup pv aren't garbage collected, etcd data pv needs to be garbage collected when not needed anymore.
+
+#### Pod name to Persistent Volume Claim name mapping
+
+The pod and related PVC are mapped 1 to 1 so thet can be easily generated and deducted. I.E. if the pod is called `etcd-0000` its related PVC name will be `etcd-0000-pvc`
+
+#### Results
+
+These changes will put the basis for initial persistent volume management BUT won't fix the failure scenario described at the start of the proposal: if a majority of pods is deleted there's no logic to recreate them with their previous data.
+
+#### Proposed enhancements to Part 1 (@xiang90)
+
+- When using a PVC as volume source also set pod restart policy to `Always`
+
+With this change when the etcd container of the pod dies due to different reasons (etcd crashing, container killed, node reboots but its pods aren't evicted by the k8s node controller) it will be restarted (since Restart Policy is set to Always). Previously the pod will go in Failed state and will be removed and etcd operator will remove the etcd member and add a new member and a new pod.
+
+
+### Part 2
+
+#### Changes
+
+- Change reconcile logic to handle deleted pods that had a persistent volume:
+
+If the a pod of an etcd cluster member is deleted and it had a persistent volume claim associated, instead of removing the etcd member from the etcd cluster and adding a new member just recreate it (keeping its pod name) binding the existing persistent volume claim. The pod recreation will happen only if the pod doesn't exists anymore in the API because we'll keep the same pod name. This also enable a MANDATORY property: at most once pod existence logic to avoid having in the API at the same time two pods with associated the same PVC.
+
+#### Results
+
+With this change if a majority of pods is deleted the etcd-operator will recreate them keeping their previous data, so the above failure scenario will be fixed.
+
+#### Notes
+
+From https://github.com/kubernetes/community/blob/master/contributors/design-proposals/pod-safety.md to avoid data corruptions, the node controller will only mark the partitioned node pods for termination but will not remove them from the api. A fence controller (not yet existing as a standard k8s component) or a manual intervention have to ensure that the node is down and then remove it from the API or force remove all of its pods from the API.
+
+So only when the pod is removed from the API the above logic can recreate it. If this isn't going to happen for a long time, the new etcd pod won't be recreated. The below proposed enhancements will help avoiding such cases.
```

Comment:
This is not what that proposal talks about.. Actually that proposal is not related to this proposal right now.. Remove this section?

Reply:
Given the above answers I think this is important since it explains why pods are evicted from dead nodes and why they aren't directly removed by the API but will wait for the node to come back or being removed from the API by an external fencing component.

Since when a majority of etcd pods are removed, to restore quorum, etcd-operator (part 2) has to recreate them with their existing data (PV), it's foundamental that etcd-operator, as a pod controller, behaves correctly:

* don't force delete pods with PV from the API (i.e don't use grace period = 0)
* obtain at most once pod guarantee reusing the previous pod name.

Reply:
In the design doc, we will only talk about the actions and the reasoning behind the actions. I can see that:
- Action: we need to handle failure because a pod could be failing permanently and we need to recover them.
- Reasoning: Pods could be evicted, Volume could fail permanently, node could fail permanently.

Other than that, what's the point of this paragraph? What's the problem it is trying to solve?

> Since when a majority of etcd pods are removed, to restore quorum, etcd-operator

So this is a **second** problem from the eviction problem you mentioned. Right? If so, separate them into two problem sections and discuss separately.

Reply:
I hope the 4 cases I depicted will help. Action/Reasoning should be covered by them.